src/test/java/emissary/core/InterruptibleCharSequenceTest.java

```diff
@@ -0,0 +1,104 @@
+        long interruptibleCharSequenceTotalTime = 0;
+        for (int i = 1; i <= trials; i++) {
+            logger.info("trial {} of {}", i, trials);
+            long start = System.currentTimeMillis();
+            tryMatch(inputNormal, regex);
+            long end = System.currentTimeMillis();
+            logger.info("String {}", end - start);
+            stringTotalTime += end - start;
+            start = System.currentTimeMillis();
+            tryMatch(inputWrapped, regex);
+            end = System.currentTimeMillis();
+            logger.info("InterruptibleCharSequence {}", end - start);
+            interruptibleCharSequenceTotalTime += end - start;
+        }
+        float overheadPercentage = (float) interruptibleCharSequenceTotalTime / stringTotalTime;
+        logger.info("InterruptibleCharSequence overhead percentage: {}.", overheadPercentage);
+    }
+
+    public boolean tryMatch(CharSequence input, String regex) {
+        return Pattern.matches(regex, input);
+    }
+
+    public Thread tryMatchInThread(final CharSequence input, final String regex, final BlockingQueue<Object> atFinish) {
+        Thread t = new Thread(() -> {
+            boolean matcherResult;
+            try {
+                matcherResult = tryMatch(input, regex);
+            } catch (InterruptedByTimeoutException interruptedByTimeoutException) {
+                atFinish.offer(interruptedByTimeoutException);
+                return;
+            }
+            atFinish.offer(matcherResult);
+        });
+        t.start();
+        return t;
+    }
+
+    @Test
+    void testNoninterruptibleString() throws InterruptedException {
+        BlockingQueue<Object> blockingQueue = new LinkedBlockingQueue<>();
+        Thread t = tryMatchInThread(INPUT, BACKTRACKER, blockingQueue);
+        sleep(1000);
+        t.interrupt();
+        Object result = blockingQueue.take();
+        assertEquals(Boolean.FALSE, result, "expected to not find a match");
+    }
+
+    @Test
+    void testInterruptibleCharSequence() throws InterruptedException {
+        long sleepMillis = 32;
+
+        while (sleepMillis > 0) {
+            BlockingQueue<Object> blockingQueue = new LinkedBlockingQueue<>();
+            Thread t = tryMatchInThread(new InterruptibleCharSequence(INPUT), BACKTRACKER, blockingQueue);
+            sleep(sleepMillis);
+            if (Thread.State.TERMINATED == t.getState()) {
+                sleepMillis /= 2;
+                logger.info("Task completed, retrying with shorter sleep time: {} ms", sleepMillis);
+                continue;
+            }
+            t.interrupt();
+            Object result = blockingQueue.take();
+
+            // Assert that InterruptedByTimeoutException was thrown
+            if (result instanceof InterruptedByTimeoutException) {
+                return;
+            } else {
+                fail("failed to interrupt InterruptibleCharSequence with given sleeping intervals");
+            }
```

Comment:
Simplify?
```suggestion
            assertTrue(result instanceof InterruptedByTimeoutException, "failed to interrupt InterruptibleCharSequence with given sleeping intervals");
```

Reply:
The while loop needs to exit or the next iteration will return false. I struggled with making this code more concise and open to other suggestions.

Reply:
Maybe?  
```java
   assertTrue(... same stuff);
   return;
```

Reply:
It's caused the PMD verify to fail the build. Didn't like `return` being the last statement in the loop.